src/rulesets/Base/mapreduce.jl

```diff
@@ -417,137 +417,190 @@ end
+##### `foldl(f, ::Array)`
 #####
 
-# This zoo of underscore functions helps `foldl` & `accumulate` handle both tuples and arrays,
-# and also provides some alternatives for versions of Julia where iterators weren't supported.
-# Inspired by `Base._reverse`, used in defn of `foldr`.
+# The implementation was originally for both tuples and arrays, although using accumulate
+# to carry intermediate results along creates arrays of tuples which could be avoided.
+# Using a loop can be a few times faster, this should be replaced:
+# https://github.com/FluxML/Zygote.jl/issues/644#issuecomment-628762305
 
-# To support 2nd derivatives, some may need their own gradient rules. And _drop1 should perhaps
-# be replaced by _peel1 like Iterators.peel
+# Note also that it does not return a gradient for `init`, now marked `@not_implemented`.
 
-_reverse1(x) = Iterators.reverse(x)
-_drop1(x) = Iterators.drop(x, 1)
-_zip2(x, y) = zip(x, y)  # for `accumulate`, below
-
-_reverse1(x::Tuple) = reverse(x)
-_drop1(x::Tuple) = Base.tail(x)
-_zip2(x::Tuple{Vararg{Any,N}}, y::Tuple{Vararg{Any,N}}) where N = ntuple(i -> (x[i],y[i]), N)
-
-struct _InitialValue end  # Old versions don't have `Base._InitialValue`
+function rrule(
+        config::RuleConfig{>:HasReverseMode}, ::typeof(Base.mapfoldl_impl), ::typeof(identity), op::G, init, x::Union{AbstractArray, Tuple};
+    ) where {G}
+    start, list = if init === Base._InitialValue()
+        Iterators.peel(x)
+    else
+        # Case with init keyword is simpler to understand first!
+        init, x
+    end
+    hobbits = accumulate(list; init=(start, nothing)) do (a, _), b
+        c, back = rrule_via_ad(config, op, a, b)
+    end
+    y = first(last(hobbits))
+    axe = axes(x)
+    project = ProjectTo(x)
+    function unfoldl(dy)
+        trio = accumulate(Iterators.reverse(hobbits); init=(0, dy, 0)) do (_, dc, _), (_, back)
+            ds, da, db = back(dc)
+        end
+        dop = sum(first, trio)
+        dx = map(last, Iterators.reverse(trio))
+        if init === Base._InitialValue()  # `hobbits` is one short
+            dx = _vcat1(trio[end][2], dx)
+        end
+        d_init = @not_implemented "gradient for foldl does not at present include init, sorry"
+        return (NoTangent(), NoTangent(), dop, d_init, project(reshape(dx, axe)))
+    end
+    return y, unfoldl
+end
 
 _vcat1(x, ys::AbstractVector) = vcat(x, ys)
 _vcat1(x::AbstractArray, ys::AbstractVector) = vcat([x], ys)
-_vcat1(x, ys::Tuple) = (x, ys...)
-
-_reshape1(x::AbstractArray, axe) = reshape(x, axe)
-_reshape1(x::Tuple, axe) = x
-
-_no_tuple_tangent(dx::Tangent) = ChainRulesCore.backing(dx)
-_no_tuple_tangent(dx) = dx
-
 
 #####
 ##### `accumulate`
 #####
 
 # Like `foldl` this by definition works in order, so it makes sense to allow stateful `f`.
 
+# Also like `foldl`, the version with a keyword `init` can't easily be given a gradient.
+# Move it down to: `_accumulate!(op, B, A::AbstractVector, dims::Nothing, init::Nothing)`
```

Comment:
But we *don't* at present support getting back a gradient for init, except if it's `nothing` and then it doesn't matter.
So we might as well put this on `accumulate`?

I am a little uncomfortable putting rules on mutating fuctions.
Though perhaps this one is safe as we are always fully overwriting `y` and never reading it?
A comment to that effect would be good if so.

Reply:
The intention was to move both to functions with positional `init`, and this mutating function was the best option I could find in Base's dispatch.

Then it could have a gradient for `init`. I didn't get around to writing one, mostly got tired of fighting tests. But at least step 1 makes step 2 easier, it would be a small PR. And for now it returns `@not_implemented` which is better than a silent zero, in theory at least.

Reply:
I do think this is unsafe the same way that `fill!` is unsafe. Except that in practice, I think it's much less likely to cause problems, as anyone who gets to `accumulate!` has probably been trained out of hoping that mutation will work.

The originally envisaged use case was that the 2nd derivative of `foldl` would involve this `accumulate` gradient. But I don't recall whether I ever checked that.